custom-metrics-stackdriver-adapter/pkg/adapter/provider/provider.go

```diff
@@ -0,0 +1,128 @@
+	_ "k8s.io/client-go/pkg/api/install"
+
+	"github.com/GoogleCloudPlatform/k8s-stackdriver/custom-metrics-stackdriver-adapter/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-stackdriver/custom-metrics-stackdriver-adapter/pkg/provider"
+)
+
+// TODO(kawych):
+// * Support metrics for pods identified by selector - in progress.
+// * Implement ListAllMetrics - in progress.
+// * Support metrics for objects other than pod, e.i. root-scoped - depends on SD resource types.
+// * Support long responses from Stackdriver (pagination).
+
+type clock interface {
+	Now() time.Time
+}
+
+type realClock struct{}
+
+func (c realClock) Now() time.Time {
+	return time.Now()
+}
+
+// StackdriverProvider is a provider of custom metrics from Stackdriver.
+type StackdriverProvider struct {
+	restClient rest.Interface
+
+	service *stackdriver.Service
+
+	config *config.GceConfig
+
+	rateInterval time.Duration
+
+	translator *Translator
+}
+
+// NewStackdriverProvider creates a StackdriverProvider
+func NewStackdriverProvider(restClient rest.Interface, stackdriverService *stackdriver.Service, rateInterval time.Duration) provider.CustomMetricsProvider {
+	gceConf, err := config.GetGceConfig("custom.googleapis.com")
+	if err != nil {
+		glog.Fatalf("Failed to retrieve GCE config: %v", err)
+	}
+
+	return &StackdriverProvider{
+		restClient:   restClient,
+		service:      stackdriverService,
+		config:       gceConf,
+		rateInterval: rateInterval,
+		translator: &Translator{
+			service:   stackdriverService,
+			config:    gceConf,
+			reqWindow: rateInterval,
+			clock:     realClock{},
+		},
+	}
+}
+
+// GetRootScopedMetricByName queries Stackdriver for metrics identified by name and not associated
+// with any namespace. Current implementation doesn't support root scoped metrics.
+func (p *StackdriverProvider) GetRootScopedMetricByName(groupResource schema.GroupResource, name string, metricName string) (*custom_metrics.MetricValue, error) {
+	return nil, fmt.Errorf("Non namespaced metrics not supported (metric name: %s)", metricName)
+}
+
+// GetRootScopedMetricBySelector queries Stackdriver for metrics identified by selector and not
+// associated with any namespace. Current implementation doesn't support root scoped metrics.
+func (p *StackdriverProvider) GetRootScopedMetricBySelector(groupResource schema.GroupResource, selector labels.Selector, metricName string) (*custom_metrics.MetricValueList, error) {
+	return nil, fmt.Errorf("Non namespaced metrics not supported (metric name: %s)", metricName)
+}
+
+// GetNamespacedMetricByName queries Stackdriver for metrics identified by name and associated
+// with a namespace.
+func (p *StackdriverProvider) GetNamespacedMetricByName(groupResource schema.GroupResource, namespace string, name string, metricName string) (*custom_metrics.MetricValue, error) {
+	matchingObject, err := p.restClient.Get().Namespace(namespace).Resource(groupResource.Resource).Name(name).Do().Get()
+	if err != nil {
+		return nil, err
```

Comment:
important note: errors returned from the provider are expected to be Kubernetes API-compatible errors (such as status errors).  Otherwise, you'll get some weird responses on error from the servers and clients.  The boilerplate now has constructors for common error types: https://github.com/kubernetes-incubator/custom-metrics-apiserver/blob/master/pkg/provider/errors.go.  See here for an example of using them: https://github.com/DirectXMan12/k8s-prometheus-adapter/blob/58a9769eaabceabb060ca9c78d274de176ae6777/pkg/custom-provider/provider.go#L108.  Right above it should be an example of throwing an internal error.